data-generator/generate_data.py

```diff
@@ -49,15 +65,94 @@ def make_changes(num_changes, vcs, event_timespan):
     if vcs == "gitlab":
         event = {
             "object_kind": "push",
+            "before": before,
             "checkout_sha": head_commit["id"],
             "commits": changes,
         }
-    if vcs == "github":
-        event = {"head_commit": head_commit, "commits": changes}
+    elif vcs == "github":
+        event = {
+            "head_commit": head_commit,
+            "before": before,
+            "commits": changes
+        }
+    else:
+        raise ValueError("Version Control System options limited to github or gitlab.")
 
     return event
 
 
+def make_all_changesets(num_events: int, vcs: str, event_timespan: int, num_changes: int = None) -> list[dict]:
+    """Make a lit of changesets of length ``num_event``
+
+    Args:
+        num_events (int): the number of changesets to generate
+        vcs: the version control system being used (options include github or gitlab
+        event_timespan: time duration (in seconds) of timestamps of generated events
+        num_changes: number of changes per changeset, defaults to a random uniform distribution
+            between 1 and 5 (optional)
+
+    Returns:
+        all_changesets: a list of dictionaries of all created changesets
+
+    """
+    all_changesets = []
+    prev_change_sha = secrets.token_hex(20)  # set a random prev sha
+    for _ in range(num_events):
+        if not num_changes:
+            num_changes = random.randrange(1, 5)
+        changeset = make_changes(
+            num_changes,
+            vcs,
+            event_timespan,
+            before=prev_change_sha
+        )
+        prev_change_sha = changeset.get("checkout_sha") or changeset.get("head_commit", {}).get("id")
+        all_changesets.append(changeset)
+
+    return all_changesets
+
+
+def make_ind_changes_from_changeset(changeset, vcs):
```

Comment:
Absolute complete nitpick - I wonder if there is a more succinct name?  like make_change ?

Reply:
Yeah, I can change that. To stop it from being confused with the already `make_changes`, maybe `make_ind_change`? or `make_single_change`? Also I realized that since I almost always use type hints, but since this project doesn't, I tried to not use them, but I used them one one function, but not other ones, which feels really weird. So I will remove them on that function.